crates/sui-core/src/transaction_driver/effects_certifier.rs

```diff
@@ -353,34 +295,37 @@ impl EffectsCertifier {
             let expired_weight = expired_aggregator.total_votes();
             let total_weight = executed_weight + rejected_weight + expired_weight;
 
+            // Quorum threshold is used here to gather as many responses as possible for summary,
+            // while making sure the loop can exit with f malicious peers.
             if total_weight >= committee.quorum_threshold() {
-                // Abort as early as possible because there is no guarantee that another response will be received.
-                if rejected_weight + expired_weight >= committee.validity_threshold() {
-                    return Err(TransactionDriverError::TransactionRejectedOrExpired(
-                        rejected_errors.join(", "),
-                        expired_errors.join(", "),
-                    ));
-                }
-                // Check if quorum can still be reached with remaining responses.
-                let remaining_weight = committee.total_votes().saturating_sub(total_weight);
-                let quorum_feasible = effects_digest_aggregators.values().any(|agg| {
-                    agg.total_votes() + remaining_weight >= committee.quorum_threshold()
-                });
-                if !quorum_feasible {
-                    break;
-                }
-            } else {
-                // Abort less eagerly for clearer error message.
-                // More responses are available when the network is live.
+                // TransactionRejected status may or may not be retriable.
                 if rejected_weight >= committee.validity_threshold() {
                     return Err(TransactionDriverError::TransactionRejected(
-                        rejected_errors.join(", "),
+                        rejected_weight,
+                        rejected_aggregator
+                            .statuses()
+                            .iter()
+                            .map(|(n, s)| format!("{}: {:?}", n.concise(), s))
+                            .join(", "),
+                    ));
+                }
+                // TransactionRejectedOrExpired status is always retriable.
```

Comment:
Do you think we need to be more specific here that retriable specifically means resubmission at a new consensus position and not retriable for wait_for_acknowledgments

Reply:
Yes! I would say most of the "interesting" retries, based on SuiError and internal logic, are submission retries. Effects query retries are only limited to RPC errors.